src/Frontend/Core.hs

```diff
@@ -76,16 +78,20 @@ class RedirectOf p where
 
 -- | Wrap anything that has 'ToHtml' and wrap it in an HTML body with complete page.
 data Frame body = Frame User body | PublicFrame body
+  deriving (Functor)
 
-makeFrame :: Page p => p -> Frame p
+makeFrame :: (ActionPersist m, ActionUserHandler m, Page p) => p -> m (Frame p)
 makeFrame p
-  | isPrivatePage p = Frame frameUserHack p
-  | otherwise       = PublicFrame p
+  | isPrivatePage p = flip Frame p <$> currentUser
+  | otherwise       = return $ PublicFrame p
 
 instance (ToHtml body) => ToHtml (Frame body) where
-    toHtmlRaw                = toHtml
-    toHtml (Frame usr bdy)   = pageFrame (Just usr) (toHtml bdy)
-    toHtml (PublicFrame bdy) = pageFrame Nothing (toHtml bdy)
+    toHtmlRaw = toHtml
+    toHtml = renderPageFrame toHtml
+
+renderPageFrame :: Monad m => (t -> HtmlT m a) -> Frame t -> HtmlT m ()
```

Comment:
It was necessary to create this functions, becuase I wanted to use ToHtml in the redirectFormHandler but got:

````
Could not deduce (ToHtml (HtmlT Identity ()))
````